Separated out external and internal interfaces for Transporter and Dispatcher

jzila opened the pull request:
Signed-off-by: John Zila john@jzila.com

jzila committed:
Separated out external and internal interfaces for Transporter and Dispatcher

Signed-off-by: John Zila <john@jzila.com>

maxtaco commented on rpc2/client.go:
```diff
@@ -11,6 +11,7 @@ func NewClient(xp Transporter, f UnwrapErrorFunc) *Client {
 
 func (c *Client) Call(method string, arg interface{}, res interface{}) (err error) {
 	var d Dispatcher
+	c.xp.Run(true)
```
why is this now needed?

jzila replied:
`Transport.GetDispatcher` no longer calls `Transport.run`, so we need to call it explicitly.

maxtaco commented on rpc2/dispatch.go:
```diff
@@ -6,9 +6,13 @@ type DecodeNext func(interface{}) error
+}
+
+type dispatcher interface {
+	Dispatcher
```
`Dispatcher` doesn't need to be exported out of the package at all.

jzila replied:
We can wrap the `Server` and `Client` functionality directly into `Transporter`, in which case dispatcher can be package local. Otherwise, `Transporter.GetDispatcher` is needed (even though it's only used inside rpc2's `Server` and `Client` constructs).

My preference would be to completely include the `Server` and `Client` APIs in `Transporter`, which would then look like:

``` go
type Transporter interface {
    IsConnected() bool
    RegisterProtocol(Protocol) error
    RunCall(method string, arg interface{}, res interface{}, f UnwrapErrorFunc) error
    RunServer() error
}
```

This way the `Transporter` only includes functionality that has external relevance, and `Client`/`Server` then only serve the purpose of providing the correct `wrap`/`unwrap` functions.

WDYT @maxtaco?

maxtaco replied:
I think this proposal misses a key point that I really want to preserve --- that `Client` and `Server` are both bound to a particular protocol. I really want to keep this.  So in other words, I think we need to export `Client` and `Server`.  The other stuff (`Packterizer` and `Dispatch`, `Message`) or OK to hide if possible.

jzila replied:
No worries. I'll keep the Transporter mostly the same as before.

maxtaco commented:
@jzila, can we move this PR over to keybase/go-framed-msgpack-rpc?

jzila commented:
@maxtaco I'll reopen it on the other repo once you're happy with the changes :)

jzila committed:
PR feedback implemented for factoring out exported interfaces

Signed-off-by: John Zila <john@jzila.com>

jzila commented:
I've implemented your feedback regarding `dispatcher`. These changes are still completely compatible with the way `keybase/client` consumes `rpc2`. I've pushed some changes to my own fork of `client` that replaces `*rpc2.Transport` with `rpc2.Transporter`, but those can wait.

maxtaco committed:
Merge pull request #1 from jzila/master

Separated out external and internal interfaces for Transporter and Dispatcher

jzila committed:
Made more internal interfaces unexported

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Refactored transport to make its behavior more clear, mostly switching
from mutex-based synchronization to channel-based synchronization.

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback for transport refactor.

Signed-off-by: John Zila <john@jzila.com>

maxtaco committed:
Support for the Notify() method

Added a small test case too

maxtaco committed:
comments

maxtaco committed:
Merge pull request #4 from keybase/issue_3

Support for the Notify() method

maxtaco committed:
namespace rename --

1. Get rid of the original RPC library
2. Promote rpc2/* to top-level directory
3. Rename rpc2 -> rpc

maxtaco committed:
Merge pull request #5 from keybase/namespace-reorg-1

Namespace reorg

jzila committed:
PR feedback by @akalin-keybase implemented

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Removed the dependency of dispatch on transport

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Transporter shouldn't care if it's run in the background or not, moved that to the server

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
A bit more cleanup on the transport loops to remove duplicated code

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #2 from jzila/transport-chan

Refactored transport synchonization from mutexes to channels

jzila committed:
Factored out decoding and encoding into codec and moved most functionality
out of message into dispatch

jzila committed:
Factored out DecodeNext and decoupled transport from message

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Implemented PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Factored out the various request types so the dispatcher can call them with a common API

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Removed the dependency of packetizer on transporter

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback implemented to consolidate constants

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #8 from jzila/transport-chan

More refactoring to normalize dependencies

jzila committed:
Refactored calls mutex to a call goroutine

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback for call locking and fixed a presumed race too

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Added EOF Handler

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Exported TypeError for autogen

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Implemented cancelation

Signed-off-by: John Zila <john@jzila.com>

maxtaco committed:
Merge pull request #9 from jzila/call-goroutine

Refactored calls mutex to a call goroutine

maxtaco committed:
Merge pull request #10 from jzila/eof-handler

Implemented EOF handler and TypeError

jzila committed:
Implemented cancel test and factored out protocol tests

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback removing time dependence in testing context cancel

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Moved tests around and updated rpc_test to use testify/assert

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #12 from jzila/ctx

Implemented cancelation and factored out tests

jzila committed:
Added dispatching of cancel requests

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Server side cancel implemented

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Separated out the dispatcher from the receiver

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Refactored tests to be in the same package, and started improving coverage

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Improved the mock codec to include encoding and reading bytes, so it can now drop in to replace the real codec

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Added dispatch test to improve coverage

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Implemented receiver tests, greatly improving coverage

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Added util test

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
No need for sleep in TestNotify because requests will be causally ordered

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Now closing servers as part of tests

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Refactored cancelation to be synchronous, giving control over the response.

Signed-off-by: John Zila <john@jzila.com>

akalin committed:
Implement PR feedback for #14

akalin committed:
Implement PR feedback

See https://github.com/keybase/client/pull/1207 .

patrickxb committed:
Downgrade New connection/EOF to Debug

jzila committed:
Added codec test to verify whether the reader is advanced in the real codec

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Changed mock codec so it doesn't advance the reader if the decode fails

Fixes #19

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Logged Encode error in handleCall. Fixes #17

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Made cancellation asynchronous. Fixes #18

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback for async cancellation. Fixes #23

Signed-off-by: John Zila <john@jzila.com>

cjb committed:
Trigger a Travis build

cjb committed:
Add minimal .travis.yml

jzila committed:
Fixed the receiver data races

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Fixed the final test races due to notify

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #24 from jzila/async-cancel

Made cancellation asynchronous. CR by @akalin-keybase

akalin committed:
Fix RPC cancellation hang bug

Otherwise, if an RPC is cancelled before the server
result is returned returned, this will cause a hang.

Also add tests.

akalin committed:
Use channel instead of sync.Once

This fixes a goroutine leak.

This closes #35. This closes #36.

akalin committed:
Fix stupid bug with previous commit

Should do RunAsync() when bg is true, not false.

jzila committed:
Fix for #47: do not block calls once stopCh is closed

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
WIP commit

jzila committed:
Added regression test for fix when calls are attempted for a closed transport

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #48 from jzila/devel

Do not block calls once stopCh is closed

jzila committed:
Finished first pass at RPC refactor

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Almost done RPC refactor.

Protocol test passing, some bugs still remaining, need to reimplement receiver test.

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Switched to using interface for communicating RPC messages

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Moved some code around to clean it up a bit

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
ProtocolTest passing, but there are errors due to the connection closing

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Made encoding async and fixed protocol test

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
cleanup and TODOs, as well as fix for issue #29

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Brought back error wrapping

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Changed rpc message size validation to have min/max instead of exact, for backwards compatibility

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
few minor fixes

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
dispatch and message tests passing

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
reintroduced receiver test

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Moved the goroutine for serving requests to one place

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Added codec and context changes to allow RPC debug tags

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Removed max length restriction to allow for backward compatibility

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Implemented feedback for first PR pass

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Implemented debug tags

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
First commit for second PR pass

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Second commit for PR round 2

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
formatting fixes

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Removed encoder pool and benchmarking code

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Final PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Fixed message test to include handling context

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
One more quick cleanup in tests

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #50 from keybase/jzila/rpc-refactor

RPC Async Refactor

jzila committed:
Fixed nil result error

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Clarified error output for decoding

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Made it possible to decode nil args

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Added another test to ensure nil errors could be decoded

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Fixed handling nil errors

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #53 from jzila/jzila/rpc-ci-fixes

Fixes for CI bugs

jzila committed:
Fixed issue with not enough data getting read

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #54 from jzila/jzila/rpc-refactor-fix

Fixed issue with not enough data getting read

jzila committed:
Don't dispatch call if the dispatcher has been closed

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Make the write channel buffered to prevent starvation

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Switched to using context.Canceled error

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Fixed cleanup code

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Added test for blocking cancel

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Fixed async cancel issue

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
If an encode gets blocked, goroutine it

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Removed blockingBuffer

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Renamed Encode to be clearer about its purpose

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Added fully async encoding method for cancels

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR Feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Made some errors recoverable

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #55 from keybase/jzila/async-fixes

Fixup async cancellation

jzila committed:
Cleaned up the error handling a bit

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Refactored for PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
More PR feedback

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #56 from jzila/jzila/recover-errors

Made some errors recoverable

jzila committed:
Fixed the data race in the protocol test

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Removed RPCErrorWrapper

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Reset the field number when we reset the bytes

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #58 from keybase/jzila/fix-decoder-logging

Reset the field number when we reset the bytes

jzila committed:
Added more test details and protected the Response from panicking

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #59 from keybase/jzila/log-decode-details

Added more test details and protected the Response from panicking

jzila committed:
Fixed method and protocol missing errors--they are no longer fatal to the transport

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Actually make the transport not die

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Added test to ensure that the transport is not terminated when a missing method occurs

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #61 from keybase/jzila/fix-rpc-nonfatal-errors

Fix RPC nonfatal errors

jzila committed:
Merge pull request #52 from keybase/jzila/debug-tags

Debug Tags Implemented

akalin committed:
Fix hang for cancelled calls

Buffer the call response channel for one
response.

This closes #62.

jzila committed:
Only send debug tags if they're present

Signed-off-by: John Zila <john@jzila.com>

maxtaco committed:
Merge pull request #64 from keybase/jzila/no-tags-if-empty

Only send debug tags if they're present

jzila committed:
Wrap errors even if the protocol or method isn't found

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
We weren't actually wrapping the decode error

Signed-off-by: John Zila <john@jzila.com>

jzila committed:
Merge pull request #65 from keybase/jzila/no-tags-if-empty

Wrap errors even if the protocol or method isn't found

jzila committed:
Cancel running calls when receiver is closed

KBFS-803 #done #time 2h

Signed-off-by: John Zila <john@jzila.com>

maxtaco committed:
factor the GenericClient interface out of autogen code and into the library

maxtaco committed:
Merge pull request #67 from keybase/generic-client

factor the GenericClient interface out of autogen code...

jzila committed:
Merge pull request #66 from keybase/jzila/receiver-close-cancel

Cancel running calls when receiver is closed
